[FSDP][optim_state_dict] The correct way to initialize optimizer states if the corresponding param is empty

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104765

When using KeyedOptimizer.init_state(), some optimizers initializes the states even if the param is empty (size() == 0) while some optimizer avoid initializing the states. There is no way FSDP can tell. Instead, FSDP should look up `optim.state`. Fortunatelly, `optim.state` does not rely on FQNs which some internal users change the FQNs.

Differential Revision: [D47285562](https://our.internmc.facebook.com/intern/diff/D47285562/)

fegin committed:
[FSDP][optim_state_dict] The correct way to initialize optimizer states if the corresponding param is empty

When using KeyedOptimizer.init_state(), some optimizers initializes the states even if the param is empty (size() == 0) while some optimizer avoid initializing the states. There is no way FSDP can tell. Instead, FSDP should look up `optim.state`. Fortunatelly, `optim.state` does not rely on FQNs which some internal users change the FQNs.

Differential Revision: [D47285562](https://our.internmc.facebook.com/intern/diff/D47285562/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] The correct way to initialize optimizer states if the corresponding param is empty"


When using KeyedOptimizer.init_state(), some optimizers initializes the states even if the param is empty (size() == 0) while some optimizer avoid initializing the states. There is no way FSDP can tell. Instead, FSDP should look up `optim.state`. Fortunatelly, `optim.state` does not rely on FQNs which some internal users change the FQNs.

Differential Revision: [D47285562](https://our.internmc.facebook.com/intern/diff/D47285562/)

[ghstack-poisoned]

fduwjj commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -476,16 +476,7 @@ def _flatten_optim_state_dict(
-                            unflat_osd_state[fqn],
-                        )
-                    )
+                    flat_osd_state[key] = copy.deepcopy(optim.state[param])
```
Does this mean the fake init has been already called here?

fegin replied:
Yes.

fegin committed:
Update on "[FSDP][optim_state_dict] The correct way to initialize optimizer states if the corresponding param is empty"


When using KeyedOptimizer.init_state(), some optimizers initializes the states even if the param is empty (size() == 0) while some optimizer avoid initializing the states. There is no way FSDP can tell. Instead, FSDP should look up `optim.state`. Fortunatelly, `optim.state` does not rely on FQNs which some internal users change the FQNs.

Differential Revision: [D47285562](https://our.internmc.facebook.com/intern/diff/D47285562/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge -f "The failing test is not related"
